react-paw-mailmerge/src/components/WriteTemplate.tsx

```diff
@@ -1,40 +1,86 @@
-import React from "react"
+import React, { useState, useRef } from "react"
 import Form from 'react-bootstrap/Form';
 import Button from "react-bootstrap/Button";
+import { CsvResult, EXTRA_COLUMNS } from "../Helpers/CsvFunctions";
 
 interface WriteTemplateProps {
-    parsedData: object[] | null;
+    parsedData: CsvResult | null;
     template: string,
     setTemplate: React.Dispatch<React.SetStateAction<string>>;
 }
 
 const WriteTemplate: React.FC<WriteTemplateProps> = ({parsedData, template, setTemplate}) => {
+    const [editingEnabled, setEditingEnabled] = useState<boolean>(true); // enable / disable form editing
+    const [currentInput, setCurrentInput] = useState<string>('');
+    const [previousInput, setPreviousInput] = useState<string>('');  // when form changes are discarded, restore to previousInput
+    const templateRef = useRef<HTMLTextAreaElement>(null);
+    const cursorRef = useRef<number | null>(null); // tracks the cursor location
 
     const handleTemplateChange = (event: React.ChangeEvent<HTMLInputElement>) => {
-        setTemplate(event.target.value);
+        setCurrentInput(event.target.value);
+        cursorRef.current = event.target.selectionStart;
+    }
+
+    const handleTemplateClick = ()=>{
+        const cursorPos = templateRef.current?.selectionStart ?? cursorRef.current ?? 0;
+        cursorRef.current = cursorPos;
     }
 
     const handleTemplateSubmit = () => {
-        console.log("Submitting template:", template)
+        console.log("Submitting template:", currentInput);
+        setPreviousInput(currentInput);
+        setTemplate(currentInput);  // state lifting
+        setEditingEnabled(false);
+    }
+
+    const handleTemplateDiscardChanges = () => {
+        setCurrentInput(previousInput);
     }
 
+    const addStringToTemplate = (field:string) => {
```

Comment:
There's a small problem where `addStringToTemplate` will modify the value of the text input even when `editingEnabled` is `false`. The new value doesn't get lifted to `App.tsx` until the user clicks on edit then save again, so the effect on the app's overall behavior is minimal.

Reply:
Nice find! Will have to check that before acting on the button press.